hudi-common/src/main/java/org/apache/hudi/avro/BinaryIndexedRecord.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.avro;
+
+import org.apache.hudi.exception.HoodieException;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.IndexedRecord;
+
+import java.io.IOException;
+
+/**
+ * An implementation of {@link IndexedRecord} which contains serialized bytes of an {@link IndexedRecord}
+ * only, and it's used for buffering in {@code ExternalSpillableMap}, and should be deserialized to object
+ * based {@link IndexedRecord} for merging or iterating.
+ */
+public class BinaryIndexedRecord implements IndexedRecord {
```

Comment:
I think the issue is that the serializer includes the full schema. Instead of creating a custom IndexedRecord, can this be fixed at the serializer level?

Reply:
Fix at serializer level has been considered, but it cannot solve the problem that Avro record buffered in memoryMap of `ExternalSpillableMap` occupied much more memory than legacy log scanner (5x for MDT payload). So just as other engines, Avro record should be transformed into binary format before put into record buffer.

Reply:
The key point now is the in-memory representation should also be binary format to decrease spilling probability, avoiding potential performance regression.

Reply:
Doesn't this incur extra conversion though resulting in a performance regression if there is no need to spill to disk? This seems like it is not addressing underlying issue of the estimator and serializer.

Reply:
sinked offline, we will add customized size estimator and serializer for avro records to keep in line with the legacy log scanner behavior first.